ARTstor.js

```diff
@@ -1,131 +1,484 @@
-			ZU.cleanAuthor(content, "artist", artist.indexOf(',') != -1),
-			'creators'
-		);
-	}
-	
-	if (data.Date) {
-		content = data.Date.replace(/^Photographed (?:in )?/, '');
-		associateData(newItem, content, "date");
-	}
-	
-	if (data.Rights) {
-		var m = data.Rights.match(/^.*(©.*)$/m);
-		if (m) associateData(newItem, m[1], "rights");
-	}
-	
-	if (data.Subject) {
-		content = data.Subject.split(/\s*;\s*/);
-		associateData(newItem, content, 'tags');
-	}
-	
-   //these might not be complete - it's pretty straightforward to add more
-	associateData(newItem, data.Title, "title");
-	associateData(newItem, data.Measurements, "artworkSize");
-	associateData(newItem, data.Material, "artworkMedium");
-	associateData(newItem, data.Repository, "archive");
-	associateData(newItem, data['ID Number'], "archiveLocation");
-	associateData(newItem, data.Description, "abstractNote");
-	
-	newItem.complete();
-}
\ No newline at end of file
+/**
+	Get the data from DOM and added it to candidateItems.
+	If the item is selected, also adds it to the selectItems.
+**/
+function getDomData(doc, candidateItems, selectItems, zinfoName, selectName) {
+    var zinfoDom = doc.getElementById(zinfoName);
+    var selectDom = doc.getElementById(selectName);
+    var ztitle = zinfoDom.getAttribute("ztitle");
+    var zid = zinfoDom.getAttribute("zid");
+    var ztid = zinfoDom.getAttribute("ztid");
+
+    var key = zid + ":" + ztid;
+    candidateItems[key] = htmlDecode(doc, ztitle);
+    if (selectDom.className.indexOf("thumbNailImageSelected") > -1) {
+        // The item is selected.
+        selectItems[key] = htmlDecode(doc, ztitle);
+    }
+}
+
+/**
+	This procedure gets the image id and type from the DOM, added it to 
+	the list and sends the list to processor.
+**/
+function doImageViewer(doc, url) {
+    var objID = doc.getElementById("objID");
+    if (objID != null) {
+        var objItems = [];
+        var objItem = objID.title;
+        // Get the image id and object type from the title attribute.
+        // This contains the objId and object type separate by : as in "AWSS35953_35953_25701160:11"
+        objItems.push(objItem);
+        processObjects(doc, url, objItems);
+    }
+}
+
+/**
+	Process the metadata window data by getting the ID from the 
+	window DOM and retries data with extra calls.
+**/
+function doMetadataWindow(doc, url) {
+    // get object id from metadata window.
+    var metaWindows = doc.getElementsByClassName('MetaDataWidgetRoot');
+    var objItems = [];
+    for (var i = 0; i < metaWindows.length; i++) {
+        // the dom id is in the form "mdwSS7730455_7730455_8806769" 
+        // that is object id prefixed with mdw.
+        var id = metaWindows[i].id.substring(3);
+        objItems.push(id + ":10"); // default type to image
+    }
+    processObjects(doc, url, objItems);
+}
+
+/**
+	This functions removes extra format tag from string and also decodes the
+	html entity string.
+**/
+function htmlDecode(doc, input) {
```

Comment:
Zotero has `ZU.unescapeHTML` https://github.com/zotero/zotero/blob/master/chrome/content/zotero/xpcom/utilities.js#L486 -- that should do most of this (except for the first bit). Don't duplicate.